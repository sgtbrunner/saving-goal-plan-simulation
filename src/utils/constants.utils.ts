// OPERATIONAL CONSTANTS
export const FIRST_STEP = 1;
export const LOCALE = 'en';
export const INITIAL_GOAL_AMOUNT = 25000;

// GENERAL CONSTANTS
export const ALT = 'alt';
export const BANNER = 'banner';
export const BUTTON = 'button';
export const ID = 'id';
export const IMG = 'img';
export const ORIGIN = 'Origin';
export const LEFT = 'left';
export const RIGHT = 'right';
export const TEST = 'test';
export const WHITE = 'white';
