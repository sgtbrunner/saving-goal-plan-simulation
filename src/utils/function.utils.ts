export const clearInvalidCharacters = (value: string): string =>
  value.replace(/\D|^0+/g, '');

export const convertInputToNumber = (value: string): number => {
  if (value.includes('.')) {
    const [decimalPart, fractionalPart] = value.split('.');
    return Number(
      [
        clearInvalidCharacters(decimalPart),
        '.',
        clearInvalidCharacters(fractionalPart),
      ].join('')
    );
  }

  return Number(value.replace(/\D|^0+/g, ''));
};
