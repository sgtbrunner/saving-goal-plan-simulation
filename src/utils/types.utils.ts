export type MonthPickerInputState = {
  step: number;
  year: string;
  month: string;
};

export type SavingGoalStoreProps = {
  goalAmount: number;
  reachDateMonth: string;
  reachDateYear: string;
  monthlyAmount: number;
  monthlyDeposits: number;
};

export type StoreState = {
  savingGoal: SavingGoalStoreProps;
};
