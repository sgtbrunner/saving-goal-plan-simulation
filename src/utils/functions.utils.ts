import { FIRST_STEP, LOCALE } from './constants.utils';
import { MonthPickerInputState } from './types.utils';

export const getDate = (customDate?: string | number): Date =>
  customDate ? new Date(customDate) : new Date();

export const monthFormatter = new Intl.DateTimeFormat(LOCALE, {
  month: 'long',
});

export const yearFormatter = new Intl.DateTimeFormat(LOCALE, {
  year: 'numeric',
});

export const getMonthPickerState = (
  step = FIRST_STEP,
  modifier = 0
): MonthPickerInputState => {
  const today = getDate();
  const targetMonthISO = today.setMonth(today.getMonth() + step + modifier, 1);

  return {
    step: step + modifier,
    year: yearFormatter.format(targetMonthISO),
    month: monthFormatter.format(targetMonthISO),
  };
};

export const getSignificantAndFractionalDigits = (
  value: string
): Array<string> => value.split('.');
