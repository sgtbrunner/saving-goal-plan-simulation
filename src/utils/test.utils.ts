import { INITIAL_GOAL_AMOUNT } from './constants.utils';
import { getDate, monthFormatter, yearFormatter } from './functions.utils';
import { SavingGoalState } from './types.utils';

// TAGS
export const ALT = 'alt';
export const BANNER = 'banner';
export const BUTTON = 'button';
export const IMG = 'img';
export const TEST = 'test';

// TEST IDs
export const CARD = 'card';
export const MAIN_COMPONENT = 'main-component';
export const MAIN_TITLE = 'main-title';
export const MONTHLY_AMOUNT = 'monthly-amount';
export const SUMMARY_TEXT = 'summary-text';

// FUNCTIONS
export const getDefaultSavingGoal = (): SavingGoalState => {
  const today = getDate();
  const nextMonth = today.setMonth(today.getMonth() + 1, 1);
  return {
    goalAmount: INITIAL_GOAL_AMOUNT,
    reachDateYear: yearFormatter.format(nextMonth),
    reachDateMonth: monthFormatter.format(nextMonth),
    monthlyAmount: INITIAL_GOAL_AMOUNT,
    monthlyDeposits: 1,
  };
};
