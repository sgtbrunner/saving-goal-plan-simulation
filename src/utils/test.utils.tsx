import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import rootReducer from '../redux/root-reducer';
import { INITIAL_GOAL_AMOUNT } from './constants.utils';
import { getDate, monthFormatter, yearFormatter } from './functions.utils';
import { SavingGoalState } from './types.utils';

// TAG CONSTANTS
export const ALT = 'alt';
export const BANNER = 'banner';
export const BUTTON = 'button';
export const IMG = 'img';
export const TEST = 'test';
export const AMOUNT = 'amount';

// TEST ID CONSTANTS
export const CARD = 'card';
export const MAIN_COMPONENT = 'main-component';
export const MAIN_TITLE = 'main-title';
export const MONTHLY_AMOUNT = 'monthly-amount';
export const SUMMARY_TEXT = 'summary-text';
export const CURRENCY_INPUT = 'currency-input';

// FUNCTIONS
export const getDefaultSavingGoal = (): SavingGoalState => {
  const today = getDate();
  const nextMonth = today.setMonth(today.getMonth() + 1, 1);
  return {
    goalAmount: INITIAL_GOAL_AMOUNT,
    reachDateYear: yearFormatter.format(nextMonth),
    reachDateMonth: monthFormatter.format(nextMonth),
    monthlyAmount: INITIAL_GOAL_AMOUNT,
    monthlyDeposits: 1,
  };
};

// RENDERER
type WrapperPropsType = {
  children: React.ReactNode | React.ReactNode[];
};

export const renderWithRedux = (
  component: React.ReactElement,
  store = createStore(rootReducer)
): RenderResult => {
  const ComponentWrapper = (props: WrapperPropsType) => (
    <Provider store={store}>{props.children}</Provider>
  );

  return render(component, {
    wrapper: ComponentWrapper as React.ComponentType,
  });
};
