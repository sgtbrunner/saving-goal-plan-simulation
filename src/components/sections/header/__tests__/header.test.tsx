import { screen, render, RenderResult, within } from '@testing-library/react';

import { ORIGIN } from '../../../../utils/constants.utils';
import { ALT, BANNER, IMG } from '../../../../utils/test.utils';
import Header from '../header.component';

const mountComponent = (): RenderResult => render(<Header />);

describe('Header component', () => {
  it('should match snapshot', () => {
    const { container } = mountComponent();

    expect(container).toMatchSnapshot();
  });

  it('should render Origin icon on header', () => {
    mountComponent();

    const header = screen.getByRole(BANNER);
    expect(header).toBeInTheDocument();

    const headerLogo = within(header).getByRole(IMG);
    expect(headerLogo).toBeInTheDocument();
    expect(headerLogo).toHaveProperty(ALT, ORIGIN);
  });
});
