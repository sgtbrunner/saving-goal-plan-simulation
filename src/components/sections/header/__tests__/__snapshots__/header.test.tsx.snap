// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should match snapshot 1`] = `
<div>
  <header
    class="sc-dlnjwi KHDXg"
  >
    <img
      alt="Origin"
      class="sc-hKFxyN xorWA"
      src="origin-logo.svg"
    />
  </header>
</div>
`;
